lib/plug/conn.ex

```diff
@@ -1388,6 +1366,46 @@ defmodule Plug.Conn do
   defp adapter_inform(%Conn{adapter: {adapter, payload}}, status, headers),
     do: adapter.inform(payload, status, headers)
 
+  @doc """
+  Request a protocol upgrade from the underlying HTTP adapter.
+
+  The precise semantics of an upgrade are deliberately left unspecified here in order to
+  support arbitrary upgrades, even to protocols which may not exist today. The primary intent of
+  this function is solely to allow an application to issue an upgrade request, not to manage how
+  a given protocol upgrade takes place or what APIs the application must support in order to serve
+  this updated protocol. For details in this regard, consult the documentation of the underlying
+  adapter (such a Plug.Cowboy or Bandit).
+
+  Takes an argument describing the requested upgrade (for example, `:websocket`), and an argument
+  which contains arbitrary data which the underlying adapter is expected to interpret in the
+  context of the requested upgrade.
+
+  If the upgrade is accepted by the adapter, the returned `Plug.Conn` will have a `state` of
+  `:upgraded`. This state is considered equivalently to a 'sent' state, and is subject to the same
+  limitation on subsequent mutating operations. Note that there is no guarantee or expectation
+  that the actual upgrade process is undertaken within this function; it is entirely possible that
+  the server will only do the actual upgrade later in the connection lifecycle.
```

Comment:
Do you have an example of an upgrade that is delayed like this?

Reply:
So far, all of them :). The easiest example to reason about is likely Cowboy, in that it handles the upgrade process itself, and is signalled to do so by returning a `{:cowboy_websocket, ....}` tuple from the handler function that wraps the `call/2` function. See https://github.com/elixir-plug/plug_cowboy/pull/88/files#diff-b24e9b2cfe656dcba94b216702e21de2f3909a8095b2babe494940726f29e799R35 for details of the Plug.Cowboy implementation of this, and https://github.com/mtrudel/bandit/pull/38/files#diff-d101e20286894b6a1debc54099265d0118e6edd96abca2dbcd2b7febe7db638fR26-R27 for the equivalent in Bandit.

Reply:
Great, thank you!